core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -1632,282 +1665,265 @@ object ZSTM {
       }
-      result
-    }
 
-    sealed abstract class JournalAnalysis extends Serializable with Product
-    object JournalAnalysis {
-      case object Invalid   extends JournalAnalysis
-      case object ReadOnly  extends JournalAnalysis
-      case object ReadWrite extends JournalAnalysis
-    }
-
-    /**
-     * Determines if the journal is invalid.
-     */
-    def isInvalid(journal: Journal): Boolean = !isValid(journal)
+      /**
+       * Commit all changes in the journal.
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[internal] def commit(): Unit = {
+        val it = map.valuesIterator
+        while (it.hasNext) it.next.commit()
+      }
 
-    /**
-     * Atomically collects and clears all the todos from any `TRef` that
-     * participated in the transaction.
-     */
-    def collectTodos(journal: Journal): MutableMap[TxnId, Todo] = {
-      import collection.JavaConverters._
+      /**
+       * Collects all todos and submits them to the executor.
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[internal] def completeTodos(executor: Executor)(implicit unsafe: Unsafe): Unit = {
+        val todos = collectTodos()
+        if (todos.nonEmpty) executor.submitOrThrow(() => execTodos(todos))
+      }
 
-      val allTodos  = new MutableMap[TxnId, Todo](DefaultJournalSize)
-      val emptyTodo = Map.empty[TxnId, Todo]
+      /**
+       * Executes the todos in the current thread, sequentially.
+       */
+      private[this] def execTodos(todos: Map[TxnId, Todo]): Unit = {
+        val it = todos.valuesIterator
+        while (it.hasNext) it.next.apply()
+      }
 
-      val it = journal.entrySet.iterator
-      while (it.hasNext) {
-        val tref = it.next.getValue.tref
-        val todo = tref.todo
+      /**
+       * Flag indicating whether the journal is invalid
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[ZSTM] def isInvalid: Boolean = !isValid
 
-        var loop = true
-        while (loop) {
-          val oldTodo = todo.get
+      /**
+       * Flag indicating whether the journal is valid
+       *
+       * '''NOTE''': This method MUST be invoked while we hold the lock on the
+       * journal
+       */
+      private[ZSTM] def isValid: Boolean = {
+        val it = map.valuesIterator
+        while (it.hasNext) if (!it.next().isValid) return false
+        true
+      }
 
-          loop = !todo.compareAndSet(oldTodo, emptyTodo)
+      /**
+       * Creates a function that can reset the journal.
+       */
+      private[ZSTM] def resetFn(): () => Unit = {
+        val currentNewValues = ZSTMUtils.newMutableMap[TRef[_], Any](map.size)
+        val itCapture        = map.iterator
+        while (itCapture.hasNext) {
```

Comment:
maybe this whole map initialization could be moved into the `newMutableMap` helper

Reply:
I don't think we can since the new map has a different `V` type than the original map